File: backend/app/upload_limits.py
from app.config import settings


def format_file_size(size_bytes: int) -> str:
    if size_bytes < 1024:
        return f"{size_bytes}B"
    if size_bytes < 1024 * 1024:
        return f"{size_bytes / 1024:.1f}KB"
    mb = size_bytes / (1024 * 1024)
    return f"{mb:.0f}MB" if mb.is_integer() else f"{mb:.1f}MB"


def submission_upload_limit() -> dict:
    max_bytes = settings.MAX_SUBMISSION_UPLOAD_BYTES
    return {
        "submission_max_bytes": max_bytes,
        "submission_max_label": format_file_size(max_bytes),
    }

File: frontend/src/utils/uploadLimits.js
export const DEFAULT_SUBMISSION_UPLOAD_LIMIT_BYTES = 50 * 1024 * 1024

export function formatFileSize(sizeBytes) {
  if (!Number.isFinite(sizeBytes)) return '-'
  if (sizeBytes < 1024) return `${sizeBytes}B`
  if (sizeBytes < 1024 * 1024) return `${(sizeBytes / 1024).toFixed(1)}KB`
  const mb = sizeBytes / (1024 * 1024)
  return Number.isInteger(mb) ? `${mb}MB` : `${mb.toFixed(1)}MB`
}

export async function loadSubmissionUploadLimit(api) {
  try {
    const res = await api.get('/settings/upload-limits')
    const maxBytes = Number(res.data?.submission_max_bytes) || DEFAULT_SUBMISSION_UPLOAD_LIMIT_BYTES
    return {
      maxBytes,
      label: res.data?.submission_max_label || formatFileSize(maxBytes)
    }
  } catch {
    return {
      maxBytes: DEFAULT_SUBMISSION_UPLOAD_LIMIT_BYTES,
      label: formatFileSize(DEFAULT_SUBMISSION_UPLOAD_LIMIT_BYTES)
    }
  }
}

export function totalRawFileSize(files) {
  return files.reduce((total, file) => total + (file.raw?.size || file.size || 0), 0)
}
